NameGame/NameGame/UI/UserGrid/ViewController/UserGridViewController.swift

```diff
@@ -0,0 +1,168 @@
+        guard let userRecords = viewModel.userRecords else {
+            return 0
+        }
+        return userRecords.count
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = UserRecordCell.dequeueCell(in: collectionView, with: UserGridViewController.reuseCellIdentifier, for: indexPath)
+        
+        let cachedImage = imageCache.image(for: viewModel.userRecords![indexPath.item])
+        cell.imageView.image = cachedImage
+    
+        return cell
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
+        let footerView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "mainMenuFooter", for: indexPath)
+        footerView.backgroundColor?.withAlphaComponent(0.5)
+        return footerView
+    }
+    
+    override func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
+        reloadVisibleCells()
+    }
+    
+    override func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        reloadVisibleCells()
+    }
+    
+    override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        if !decelerate {
+            reloadVisibleCells()
+        }
+    }
+
+}
+
+extension UserGridViewController {
+    
+    fileprivate func reloadVisibleCells() {
+        guard let collectionView = self.collectionView else {
+            return
+        }
+        
+        collectionView.reloadItems(at: collectionView.indexPathsForVisibleItems)
+    }
+    
+    func inject(_ imageCache: ImageCache) {
+        self.imageCache = imageCache
+    }
+    
+}
+
+extension UserGridViewController: ImageCacheEventObserver {
+    
+    func didCache(image: UIImage, for userRecord: UserRecord) {
+        guard let collectionView = self.collectionView,
+            let userRecords = viewModel.userRecords else {
+            return
+        }
+        
+        if let index = userRecords.index(of: userRecord) {
+            let indexPath = IndexPath(item: index, section: 0)
+            let cell = UserRecordCell.cell(in: collectionView, at: indexPath)
+            cell?.imageView.image = image
+        }
+    }
+}
+
+extension UserGridViewController: StoreSubscriber {
+    
+    func newState(state: UserGridViewModel) {
+        viewModel = state
+        renderLoadingStatus(with: viewModel)
+    }
+    
+    func renderLoadingStatus(with viewModel: UserGridViewModel) {
+        if viewModel.shouldShowActivityIndicator {
+            activityIndicator.startAnimating()
+        } else {
+            activityIndicator.stopAnimating()
+        }
+        
+        guard let errorMessage = viewModel.errorMessage else {
```

Comment:
Let's move this `guard` to the `presentAlert` function and change it's interface.

```swift
func presentAlert(with viewModel) {
   guard let errorMessage = ...
}
```

Then we can just call it from the `newState` function.

```swift
viewModel = state
renderLoadingStatus(with: viewModel)
renderUserRecords(with: viewModel) // <-- We are missing this one. 😉
presentAlert(with: viewModel)
```